File: ios/Sources/UnifiedTrackingPlugin/Analytics/SegmentProvider.swift
import Foundation
import Segment

/// Segment (analytics-ios v4 / analytics-swift v1) bridge for the unified-tracking plugin.
///
/// Reads `writeKey` (required) from the provider config and initializes the
/// Segment SDK. If no write key is supplied the SDK is not started and the
/// provider stays uninitialized.
///
/// NOTE(unverified): build with Xcode — confirm which Segment SDK is resolved.
/// `analytics-ios` v4 is the Objective-C SDK (uses `Analytics.setup(with:)`,
/// `Analytics.shared()`, `track:properties:` etc.) and `analytics-swift` v1+ is
/// the Swift rewrite (uses `Analytics(configuration:)`, `analytics.track(...)`).
/// The `Package.swift` declares `analytics-ios ~4.1.8` which is the ObjC SDK —
/// the API below assumes that. If the resolved package is `analytics-swift`,
/// switch to `Analytics(configuration: Configuration(writeKey:))`.
class SegmentProvider: AnalyticsProvider {
    let name = "segment"
    private(set) var isEnabled: Bool
    private(set) var isInitialized: Bool

    private let config: [String: Any]
    private var debugMode = false
    /// Tracks the consent state for parity. Segment exposes `enable()` / `disable()`.
    private var consentGranted = true
    /// The Segment `Analytics` instance. ObjC SDK exposes a shared singleton via
    /// `Analytics.shared()`; we hold a reference for explicit calls and so the
    /// provider stays uninitialized if setup fails.
    private var analytics: Analytics?

    init(config: [String: Any]) {
        self.config = config
        self.isEnabled = config["enabled"] as? Bool ?? true
        self.isInitialized = false

        guard let writeKey = config["writeKey"] as? String, !writeKey.isEmpty else {
            // No write key -> cannot start Segment. Leave uninitialized.
            print("[SegmentProvider] No writeKey provided; Segment not started.")
            return
        }

        // NOTE(unverified): build with Xcode — confirm `AnalyticsConfiguration(writeKey:)`
        // initializer and `Analytics.setup(with:)` / `Analytics.shared()` exist on
        // analytics-ios ~4.1.8. The Swift SDK uses
        // `Analytics(configuration: Configuration(writeKey:))` instead.
        let configuration = AnalyticsConfiguration(writeKey: writeKey)
        // Optional tuning hooks exposed on AnalyticsConfiguration.
        if let trackApplicationLifecycle = config["trackApplicationLifecycleEvents"] as? Bool {
            configuration.trackApplicationLifecycleEvents = trackApplicationLifecycle
        }
        if let recordScreenViews = config["recordScreenViews"] as? Bool {
            configuration.recordScreenViews = recordScreenViews
        }
        Analytics.setup(with: configuration)
        self.analytics = Analytics.shared()

        self.isInitialized = true
    }

    func trackEvent(_ event: String, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let analytics = analytics else { return }
        // NOTE(unverified): build with Xcode — confirm `track(_:properties:)` argument
        // labels in analytics-ios ~4.1.8 (ObjC bridges as `track:properties:`).
        analytics.track(event, properties: properties)
    }

    func identifyUser(_ userId: String, traits: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let analytics = analytics else { return }
        // NOTE(unverified): build with Xcode — confirm `identify(_:traits:)` argument
        // labels in analytics-ios ~4.1.8.
        analytics.identify(userId, traits: traits)
    }

    func setUserProperties(_ properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let analytics = analytics else { return }
        // Segment retains the previously-identified userId when `identify` is called
        // with `nil`; passing only traits updates the user's traits without changing
        // identity. The exact nil-handling shape depends on SDK version.
        // NOTE(unverified): build with Xcode — confirm `identify(_:traits:)` accepts
        // `nil` userId in analytics-ios ~4.1.8.
        analytics.identify(nil, traits: properties)
    }

    func logScreenView(_ screenName: String, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let analytics = analytics else { return }
        // NOTE(unverified): build with Xcode — confirm `screen(_:properties:)` argument
        // labels in analytics-ios ~4.1.8.
        analytics.screen(screenName, properties: properties)
    }

    func logRevenue(amount: Double, currency: String, productId: String?, quantity: Int, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let analytics = analytics else { return }

        var props: [String: Any] = [
            // Segment's ecommerce spec uses `revenue` as the canonical monetary key.
            "revenue": amount,
            "currency": currency,
            "quantity": quantity
        ]
        if let productId = productId {
            props["productId"] = productId
        }
        // Caller-supplied properties take precedence over the derived defaults.
        for (key, value) in properties {
            props[key] = value
        }

        // Segment ecommerce spec event name.
        analytics.track("Order Completed", properties: props)
    }

    func setConsent(_ granted: Bool) {
        consentGranted = granted
        guard let analytics = analytics else { return }
        // NOTE(unverified): build with Xcode — confirm `enable()` / `disable()` instance
        // methods exist on analytics-ios ~4.1.8. These stop/start event delivery on
        // the underlying queue.
        if granted {
            analytics.enable()
        } else {
            analytics.disable()
        }
    }

    func reset() {
        guard let analytics = analytics else { return }
        // NOTE(unverified): build with Xcode — confirm `reset()` clears anonymousId +
        // identifyUser/traits in analytics-ios ~4.1.8.
        analytics.reset()
    }

    func setDebugMode(_ enabled: Bool) {
        // Segment's debug logging is enabled via `Analytics.debug(true)` (a class
        // method) in analytics-ios. We retain the flag here for parity and call the
        // class method when possible.
        debugMode = enabled
        // NOTE(unverified): build with Xcode — confirm `Analytics.debug(_:)` class
        // method exists in analytics-ios ~4.1.8.
        Analytics.debug(enabled)
    }
}

File: android/src/main/java/com/aoneahsan/unifiedtracking/providers/analytics/SegmentProvider.java
package com.aoneahsan.unifiedtracking.providers.analytics;

import android.app.Application;
import android.content.Context;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.aoneahsan.unifiedtracking.providers.AnalyticsProvider;
import com.segment.analytics.Analytics;
import com.segment.analytics.Properties;
import com.segment.analytics.Traits;

import java.util.Map;

/**
 * Segment analytics provider for Android.
 *
 * Uses {@code com.segment.analytics.Analytics} from the Segment Analytics-Android SDK
 * (artifact {@code com.segment.analytics.android:analytics}). The SDK is built once
 * via {@code Analytics.Builder(Context, writeKey).build()} and then promoted to the
 * singleton instance via {@code Analytics.setSingletonInstance(...)} so consumer code
 * elsewhere in the app can call {@code Analytics.with(context)}.
 *
 * NOTE(unverified): built against Segment Analytics-Android SDK 4.10.x API. Confirm
 * exact signatures (Builder accepts Context or Application; Properties.putValue vs
 * putAll; Traits.putAll; Analytics.with(context).optOut(boolean); analytics.reset())
 * in Android Studio against the resolved SDK version.
 */
public class SegmentProvider implements AnalyticsProvider {
    private static final String TAG = "Segment";
    private final Context context;
    private final JSObject config;
    private Analytics analytics;
    private boolean enabled;
    private boolean initialized;
    private boolean debugMode = false;

    public SegmentProvider(Context context, JSObject config) {
        this.context = context;
        this.config = config;
        this.enabled = config.getBoolean("enabled", true);
        this.initialized = false;

        try {
            String writeKey = config.getString("writeKey");
            if (writeKey == null || writeKey.trim().isEmpty()) {
                Log.e(TAG, "Segment writeKey is missing; provider will be inert.");
                return;
            }

            // NOTE(unverified): Segment Analytics-Android Builder requires an Application
            // context. We attempt to upcast the provided Context; if it isn't an
            // Application, fall back to context.getApplicationContext() to satisfy the
            // builder's expectation. Confirm Builder signature in Android Studio.
            Application app;
            Context appCtx = context.getApplicationContext();
            if (appCtx instanceof Application) {
                app = (Application) appCtx;
            } else {
                // The Capacitor activity's context should always wrap an Application;
                // if not, the builder may throw — log and bail out cleanly.
                Log.e(TAG, "Segment requires an Application context; provider will be inert.");
                return;
            }

            this.analytics = new Analytics.Builder(app, writeKey).build();
            Analytics.setSingletonInstance(this.analytics);
            this.initialized = true;

            if (debugMode) {
                Log.d(TAG, "Segment initialized");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Segment", e);
            this.initialized = false;
        }
    }

    @Override
    public String getName() {
        return "segment";
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public boolean isInitialized() {
        return initialized;
    }

    @Override
    public void trackEvent(String event, Map<String, Object> properties) {
        if (!enabled || !initialized || analytics == null || event == null) return;

        try {
            Properties props = buildProperties(properties);
            analytics.track(event, props);

            if (debugMode) {
                Log.d(TAG, "Tracked event: " + event);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to track event", e);
        }
    }

    @Override
    public void identifyUser(String userId, Map<String, Object> traits) {
        if (!enabled || !initialized || analytics == null || userId == null) return;

        try {
            Traits t = buildTraits(traits);
            // NOTE(unverified): Analytics.identify(String, Traits, Options) is the
            // 3-arg overload in Segment Analytics-Android 4.x. Passing null Options
            // applies the default integration set. Confirm the overload exists in
            // the resolved SDK.
            analytics.identify(userId, t, null);

            if (debugMode) {
                Log.d(TAG, "Identified user: " + userId);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to identify user", e);
        }
    }

    @Override
    public void setUserProperties(Map<String, Object> properties) {
        if (!enabled || !initialized || analytics == null || properties == null) return;

        try {
            Traits t = buildTraits(properties);
            // Segment retains the previously-set userId when identify is called with
            // only traits, so we don't pass userId here. NOTE(unverified): Confirm the
            // 1-arg identify(Traits) overload exists in the resolved SDK; otherwise
            // fall back to identify(currentUserId, traits, null).
            analytics.identify(t);

            if (debugMode) {
                Log.d(TAG, "Set user properties");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set user properties", e);
        }
    }

    @Override
    public void logRevenue(double amount, String currency, String productId, int quantity, Map<String, Object> properties) {
        if (!enabled || !initialized || analytics == null) return;

        try {
            // Segment's standard "Order Completed" semantic event. NOTE(unverified):
            // Properties.putValue(String, Object) exists in v4.x; alternative APIs
            // include putRevenue/putCurrency/putValue. We use putValue uniformly so
            // the call works against any 4.x release.
            Properties props = new Properties();
            props.putValue("revenue", amount);
            if (currency != null) props.putValue("currency", currency);
            if (productId != null) props.putValue("productId", productId);
            props.putValue("quantity", quantity);

            if (properties != null) {
                for (Map.Entry<String, Object> entry : properties.entrySet()) {
                    if (entry.getKey() == null || entry.getValue() == null) continue;
                    props.putValue(entry.getKey(), entry.getValue());
                }
            }

            analytics.track("Order Completed", props);

            if (debugMode) {
                Log.d(TAG, "Logged revenue: " + amount + " " + currency);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to log revenue", e);
        }
    }

    @Override
    public void logScreenView(String screenName, Map<String, Object> properties) {
        if (!enabled || !initialized || analytics == null || screenName == null) return;

        try {
            Properties props = buildProperties(properties);
            analytics.screen(screenName, props);

            if (debugMode) {
                Log.d(TAG, "Logged screen view: " + screenName);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to log screen view", e);
        }
    }

    @Override
    public void setConsent(boolean granted) {
        if (!initialized) return;

        try {
            // NOTE(unverified): Analytics.with(context).optOut(boolean) toggles event
            // dispatch in Segment Analytics-Android v4.x. The local instance also
            // exposes optOut(...) on some SDK builds; we use the static accessor to
            // match the documented usage.
            Analytics.with(context).optOut(!granted);
            this.enabled = granted;

            if (debugMode) {
                Log.d(TAG, "Set consent: " + granted);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set consent", e);
        }
    }

    @Override
    public void reset() {
        if (!initialized || analytics == null) return;

        try {
            analytics.reset();

            if (debugMode) {
                Log.d(TAG, "Reset Segment state");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to reset", e);
        }
    }

    @Override
    public void setDebugMode(boolean enabled) {
        this.debugMode = enabled;
        if (enabled) {
            Log.d(TAG, "Debug mode enabled for Segment provider");
        }
    }

    private Properties buildProperties(Map<String, Object> map) {
        Properties props = new Properties();
        if (map == null) return props;
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            if (entry.getKey() == null || entry.getValue() == null) continue;
            props.putValue(entry.getKey(), entry.getValue());
        }
        return props;
    }

    private Traits buildTraits(Map<String, Object> map) {
        Traits traits = new Traits();
        if (map == null) return traits;
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            if (entry.getKey() == null || entry.getValue() == null) continue;
            traits.putValue(entry.getKey(), entry.getValue());
        }
        return traits;
    }
}
